### PR TITLE
Refactor `HEBOSampler` based on `TPESampler`

### PR DESCRIPTION
**Pull request by nabenabe0928**
This PR refactors `HEBOSampler` based on `TPESampler`.

### CODE REVIEW

File: package/samplers/hebo/sampler.py [hunk 1]
```diff
@@ -240,7 +237,9 @@ def sample_independent(
         param_name: str,
         param_distribution: BaseDistribution,
     ) -> Any:
-        if not self._is_independent_sample_necessary:
+        states = (TrialState.COMPLETE, TrialState.RUNNING)
+        trials = study._get_trials(deepcopy=False, states=states, use_cache=True)
+        if any(param_name in trial.params for trial in trials):
```
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:242]: This comes from `TPESampler`, c.f., [code](https://github.com/optuna/optuna/blob/467cca671e4940145f2feffe173dfcfc153b5744/optuna/samplers/_tpe/sampler.py#L414).

File: package/samplers/hebo/sampler.py [hunk 2]
```diff
@@ -240,10 +240,10 @@ def sample_independent(
         param_name: str,
         param_distribution: BaseDistribution,
     ) -> Any:
-        if not self._is_independent_sample_necessary:
-            warnings.warn(
-                "`HEBOSampler` falls back to `RandomSampler` due to dynamic search space."
-            )
+        states = (TrialState.COMPLETE, TrialState.RUNNING)
```
Comment (nabenabe0928) [inline package/samplers/hebo/sampler.py:243]:
```suggestion
        states = (TrialState.COMPLETE, )
```

**Comment by nabenabe0928** [comment]
@eukaryo 
Could you take a look at this PR?

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Pull request merged**